main.nf

```diff
@@ -0,0 +1,519 @@
+summary['Current home']     = "$HOME"
+summary['Current user']     = "$USER"
+summary['Current path']     = "$PWD"
+summary['Working dir']      = workflow.workDir
+summary['Output dir']       = params.outdir
+summary['Script dir']       = workflow.projectDir
+summary['Config Profile']   = workflow.profile
+if(workflow.profile == 'awsbatch'){
+   summary['AWS Region'] = params.awsregion
+   summary['AWS Queue'] = params.awsqueue
+}
+if(params.email) summary['E-mail Address'] = params.email
+log.info summary.collect { k,v -> "${k.padRight(15)}: $v" }.join("\n")
+log.info "========================================="
+
+
+def create_workflow_summary(summary) {
+
+    def yaml_file = workDir.resolve('workflow_summary_mqc.yaml')
+    yaml_file.text  = """
+    id: 'nf-core-deepvariant-summary'
+    description: " - this information is collected when the pipeline is started."
+    section_name: 'nf-core/deepvariant Workflow Summary'
+    section_href: 'https://github.com/nf-core/deepvariant'
+    plot_type: 'html'
+    data: |
+        <dl class=\"dl-horizontal\">
+${summary.collect { k,v -> "            <dt>$k</dt><dd><samp>${v ?: '<span style=\"color:#999999;\">N/A</a>'}</samp></dd>" }.join("\n")}
+        </dl>
+    """.stripIndent()
+
+   return yaml_file
+}
+
+
+/********************************************************************
+  preprocess fasta files processes
+  Collects all the files related to the reference genome, like
+  .fai,.gz ...
+  If the user gives them as an input, they are used
+  If not they are produced in this process given only the fasta file.
+********************************************************************/
+
+if(!fai) {
+  process preprocess_fai {
+      publishDir "$baseDir/sampleDerivatives"
+
+      input:
+      file(fasta) from fastaToIndexCh
+
+      output:
+      file("${fasta}.fai") into faiCh
+
+      script:
+      """
+      samtools faidx $fasta
+      """
+  }
+}
+
+if(!fastagz) {
+  process preprocess_fastagz {
+      publishDir "$baseDir/sampleDerivatives"
+
+      input:
+      file(fasta) from fastaToGzCh
+
+      output:
+      file("*.gz") into (tmpFastaGzCh, fastaGzCh)
+
+      script:
+      """
+      bgzip -c ${fasta} > ${fasta}.gz
+      """
+  }
+}
+
+if(!gzfai) {
+  process preprocess_gzfai {
+    publishDir "$baseDir/sampleDerivatives"
+
+    input:
+    file(fasta) from fastaToGzFaiCh
+    file(fastagz) from tmpFastaGzCh
+
+    output:
+    file("*.gz.fai") into gzFaiCh
+
+    script:
+    """
+    samtools faidx $fastagz
+    """
+  }
+}
+
+if(!gzi){
+  process preprocess_gzi {
+    publishDir "$baseDir/sampleDerivatives"
+
+    input:
+    file(fasta) from fastaToGziCh
+
+    output:
+    file("*.gz.gzi") into gziCh
+
+    script:
+    """
+    bgzip -c -i ${fasta} > ${fasta}.gz
+    """
+  }
+}
+
+//channel to collect all necessary reference files
+fastaChannel = Channel.from(fastaCh).merge(bedCh, faiCh, fastaGzCh, gzFaiCh, gziCh).collect()
```

Comment:
Even though many of the files eg `fastagz`, `gzfai` & `gzi` etc are not used in the script block they are used by `make_examples.py` therefore the `make_examples` process inputs would need to look like this:
```nextflow
input:
file fai from faiToExamples.collect()
file fastagz from fastaGzToExamples.collect()
file gzfai from gzFaiToExamples.collect()
file gzi from gziToExamples.collect()
file bed from bedToExamples.collect()
set file(bam), file(bai) from completeChannel
```

Something I am not sure on is if the files need to be grouped together for when multiple inputs are passed with the `bam_folder` parameter

I would also need to make each of these channels twice both for the `genome` and `fasta` option. 

Do you think this would be an improvement? <br /><br />

Regarding the `fastaChannel` being used twice. Yes, I was surprised that it works despite being used twice and I'm not sure why this is.

If you would prefer that I use seperate channels, shall I do an easy fix with something like this:
```nextflow
fastaChannel = Channel.from(fastaCh)
                  .merge(bedCh, faiCh, fastaGzCh, gzFaiCh, gziCh)
                  .collect()
                  .into { fastaChannelToExample;   fastaChannelToVariants}
```

Of course, this won't be required if I use the method above with `.collect()` for both processes

Many thanks in advance

Reply:
Hi @PhilPalmer - your reply seems to have gone missing but I got it in my inbox..

> Even though many of the files eg fastagz, gzfai & gzi etc are not used in the script block they are used by `make_examples.py`

Gotcha - I didn't realise that they were implicitly required as they weren't mentioned in the command. I understand why it's built like this then 👍 I think the more verbose style is clearer to read though, so probably worth the extra lines.

> I would also need to make each of these channels twice both for the genome and fasta option.

I don't think you'd need to, when you use `.collect()` the channel isn't consumed (I think), so it can be used elsewhere.

Anyway, I guess this is all resolved now, so a moot point.